### ISSUE TITLE
Generated code of a routine with TIMESTAMP parameter has too little precision

### ISSUE THREAD

**Issue by SOERGI** [jOOQ/jOOQ#11191]
I wasn't sure if this is going to be fixed in https://github.com/jOOQ/jOOQ/issues/9945 so I created this issue. Feel free to close it if it is already cover there.

### Expected and actual behavior

I have a routine with the following signature:
```
CREATE OR REPLACE FUNCTION get_updated_person(last_updated_timestamp TIMESTAMP)
    RETURNS TABLE(person_id INT, ...)
    LANGUAGE plpgsql
    SET default_tablespace TO ''
AS
$$
BEGIN
...
END
$$
```

which leads to this generated code:

```
public GetUpdatedPerson call(
          LocalDateTime lastUpdatedTimestamp
    ) {
        GetUpdatedPerson result = new GetUpdatedPerson(DSL.name("get_updated_person"), null, new Field[] {
              DSL.val(lastUpdatedTimestamp, SQLDataType.LOCALDATETIME(0))            
        });

        return aliased() ? result.as(getUnqualifiedName()) : result;
    }
```

older versions (e.g. 3.13.x) used to generate 

```
public GetUpdatedPerson call(LocalDateTime lastUpdatedTimestamp) {
        return new GetUpdatedPerson(DSL.name(getName()), null, new Field[] { 
              DSL.val(lastUpdatedTimestamp, org.jooq.impl.SQLDataType.LOCALDATETIME)
        });
    }
```

Note the `LOCALDATETIME(0)` vs `LOCALDATETIME`, which leads to a loss of precision (which is obviously needed :-) ). As a quickfix I changed the `0` to `6`.  

### Steps to reproduce the problem

The config:

```
<plugin>
  <groupId>org.jooq</groupId>
  <artifactId>jooq-codegen-maven</artifactId>

  <configuration>
      <!-- JDBC connection parameters -->
      <jdbc>
          <driver>${db.driver}</driver>
          <url>${db.url}</url>
          <user>${db.user}</user>
          <password>${db.password}</password>
      </jdbc>

      <!-- Generator parameters -->
      <generator>
          <database>
              <name>org.jooq.meta.postgres.PostgresDatabase</name>
              <includes>.*</includes>
              <excludes>
                  schema_version # flyway
                  | flyway_schema_history # flyway
              </excludes>
              <inputSchema>${db.schema}</inputSchema>
              <tableValuedFunctions>true</tableValuedFunctions>
          </database>

          <generate>
              <javaTimeTypes>true</javaTimeTypes>
          </generate>

          <target>
              <packageName>com.our.package</packageName>
              <directory>src/main/java</directory>
          </target>
      </generator>
  </configuration>
</plugin>
```
I thought the above details should be enough, but I can create a separate MCVE if needed, just let me know.

### Versions

- jOOQ: 3.14.3
- Java: 11
- Database (include vendor): PostgreSQL
- OS: MacOS
- JDBC Driver (include name if inofficial driver): org.postgresql.Driver

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report. I think an MCVE won't be needed, this should be easy to reproduce. It's not the same as #9945 - I think that generating default precisions has simply been overlooked for stored procedures. Will look into this very soon.

**Comment by lukaseder** [comment]
Analysing this now. PostgreSQL doesn't report any `information_schema.routines.datetime_precision` nor `information_schema.parameters.datetime_precision`, so we cannot reverse engineer the exact precision easily. In any case, we must avoid setting an explicit precision of zero if we don't know the precision. The value `0` probably appears because of some `(Integer) null` to `(int) 0` conversion (the primitive default value being zero)

**Comment by lukaseder** [comment]
Yes, `DefaultDataTypeDefinition` stores the precision as `int`, not `Integer`, thus the loss of information. I'll have to think about how this can be fixed. Perhaps, a quick fix is the most reasonable approach here, given that this may affect other locations, too, i.e. a flag that allows for turning off the generation of timestamp precision information.

**Comment by andbed** [comment]
Hi @lukaseder any plans to resolve this issue?
It is a show stopper for us, therefore we cannot upgrade JOOQ version since half a year ;(

**Comment by lukaseder** [comment]
> Hi @lukaseder any plans to resolve this issue?

Yes. 3.15 won't be released without this fix, and a backport will be made as soon as ready.

> It is a show stopper for us, therefore we cannot upgrade JOOQ version since half a year ;(

Don't forced types work for you as a workaround? https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types

**Comment by lukaseder** [comment]
In the long run, it makes sense to replicate the runtime meta data's `precisionDefined()` (and `lengthDefined()` / `scaleDefined()`) properties to remember whether `0` meant no precision or zero precision (#12048). I'm reluctant to backport such a significant change to the `jOOQ-meta` model to a late patch release of 3.14, so perhaps as a quicker fix, we'll just hard-wire the precision to `6`, which can be shown by this simple example:

```sql
CREATE FUNCTION f_timestamp_precision(p_ts timestamp)
RETURNS timestamp
AS $$
BEGIN
  RETURN p_ts;
END;
$$ LANGUAGE plpgsql;

select extract(microseconds from f_timestamp_precision('2000-01-01 00:00:00.123456789'::timestamp));
```

Producing the rounded precision `123457.0`

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.12 (#12049)

**Comment by lukaseder** [comment]
3.14.12 has been released including this fix
